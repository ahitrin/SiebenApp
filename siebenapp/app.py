#!/usr/bin/env python3
# coding: utf-8
import sys
from argparse import ArgumentParser
from os.path import dirname, join, realpath

from PyQt5.QtCore import pyqtSignal, Qt, QRect
from PyQt5.QtGui import QPainter, QPen
from PyQt5.QtWidgets import QApplication, QMainWindow, QWidget, QGridLayout
from PyQt5.uic import loadUi

from siebenapp.goaltree import Edge
from siebenapp.render import Renderer
from siebenapp.system import save, load, DEFAULT_DB, split_long
from siebenapp.ui.goalwidget import Ui_GoalBody


class GoalWidget(QWidget, Ui_GoalBody):
    clicked = pyqtSignal()

    def __init__(self):
        super().__init__()
        self._click_in_progress = False
        self.setupUi(self)
        self.is_real = True
        self.widget_id = None

    def setup_data(self, number, attributes):
        self.widget_id = number
        self.label_goal_name.setText(split_long(attributes['name']))
        self.check_open.setVisible(attributes['switchable'])
        self.is_real = isinstance(number, int)
        selection = attributes['select']
        if selection == 'select':
            self.setStyleSheet('background-color:#808080;')
        elif selection == 'prev':
            self.setStyleSheet('background-color:#C0C0C0;')
        if self.is_real:
            frame_color = 'red' if attributes['open'] else 'green'
            border = 2 if attributes['switchable'] else 1
            self.frame.setStyleSheet('.QFrame{ border: %dpx solid %s }' % (border, frame_color))
            self.label_number.setText(str(number))
        else:
            self.setStyleSheet('color: #EEEEEE; border: #EEEEEE')

    def mousePressEvent(self, event):                           # pylint: disable=unused-argument
        self._click_in_progress = True

    def mouseReleaseEvent(self, event):                         # pylint: disable=unused-argument
        if self._click_in_progress:
            self.clicked.emit()
        self._click_in_progress = False

    def top_point(self):
        rect = self.geometry()
        return (rect.topLeft() + rect.topRight()) / 2

    def bottom_point(self):
        rect = self.geometry()
        return (rect.bottomLeft() + rect.bottomRight()) / 2


class CentralWidget(QWidget):
    EDGE_PENS = {
        Edge.BLOCKER: QPen(Qt.black, 1, Qt.DashLine),
        Edge.PARENT: QPen(Qt.black, 1, Qt.SolidLine),
    }

    def __init__(self):
        super().__init__()
        self.setGeometry(QRect(0, 0, 576, 273))
        self.setObjectName('scrollAreaWidgetContents')
        layout = QGridLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        self.setLayout(layout)
        self.dependencies = dict()

    def setDependencies(self, new_dependencies):
        self.dependencies = new_dependencies

    def paintEvent(self, event):                                    # pylint: disable=unused-argument
        painter = QPainter(self)

        widgets = {w.widget_id: (w.top_point(), w.bottom_point(), w.is_real)
                   for w in self.children() if isinstance(w, GoalWidget)}
        for widget_id, points in widgets.items():
            line_start = points[0]
            for edge in self.dependencies[widget_id]:
                line_end = widgets[edge[0]][1]
                painter.setPen(self.EDGE_PENS[edge[1]])
                painter.drawLine(line_start, line_end)
            if not points[2]:
                style = max([e[1] for e in self.dependencies[widget_id]])
                painter.setPen(self.EDGE_PENS[style])
                painter.drawLine(points[0], points[1])


class SiebenApp(QMainWindow):
    refresh = pyqtSignal()
    quit_app = pyqtSignal()

    def __init__(self, db, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.refresh.connect(self.save_and_render)
        self.quit_app.connect(QApplication.instance().quit)
        self.db = db
        self.goals = load(db, self.show_user_message)
        self.force_refresh = True

    def setup(self):
        self.action_Hotkeys.triggered.connect(self.hotkeys.show)
        self.action_About.triggered.connect(self.about.show)
        # Re-creation of scrollAreaWidgetContents looks like dirty hack,
        # but at the current moment I haven't found a better solution.
        # Widget creation in __init__ does not work: lines disappear.
        # Also we have to disable pylint warning in order to make build green.
        self.scrollAreaWidgetContents = CentralWidget()         # pylint: disable=attribute-defined-outside-init
        # End of 'looks like dirty hack'
        self.scrollArea.setWidget(self.scrollAreaWidgetContents)
        self._update_title()
        self.refresh.emit()

    def _update_title(self):
        self.setWindowTitle('{0} ({1} view) - SiebenApp'.format(self.db, self.goals.view))

    def close_goal(self, goal_id):
        def inner():
            self.goals.select(goal_id)
            self.goals.toggle_close()
            self.refresh.emit()
        return inner

    def save_and_render(self):
        if not self.goals.events and not self.force_refresh:
            return
        self.force_refresh = False
        self.statusBar().clearMessage()
        save(self.goals, self.db)
        for child in self.scrollAreaWidgetContents.children():
            if isinstance(child, GoalWidget):
                child.deleteLater()
        graph = Renderer(self.goals).build()
        if 'setDependencies' in dir(self.scrollAreaWidgetContents):
            self.scrollAreaWidgetContents.setDependencies({g: graph[g]['edge'] for g in graph})
        for goal_id, attributes in graph.items():
            widget = GoalWidget()
            self.scrollAreaWidgetContents.layout().addWidget(widget, attributes['row'], attributes['col'])
            widget.setup_data(goal_id, attributes)
            if isinstance(goal_id, int):
                widget.clicked.connect(self.select_number(goal_id))
                widget.check_open.clicked.connect(self.close_goal(goal_id))
        self.scrollAreaWidgetContents.update()

    def keyPressEvent(self, event):
        key_handlers = {
            Qt.Key_1: self.select_number(1),
            Qt.Key_2: self.select_number(2),
            Qt.Key_3: self.select_number(3),
            Qt.Key_4: self.select_number(4),
            Qt.Key_5: self.select_number(5),
            Qt.Key_6: self.select_number(6),
            Qt.Key_7: self.select_number(7),
            Qt.Key_8: self.select_number(8),
            Qt.Key_9: self.select_number(9),
            Qt.Key_0: self.select_number(0),
            Qt.Key_A: self.start_edit('Add new goal', self.goals.add),
            Qt.Key_C: self.with_refresh(self.goals.toggle_close),
            Qt.Key_D: self.with_refresh(self.goals.delete),
            Qt.Key_I: self.start_edit('Insert new goal', self.goals.insert),
            Qt.Key_K: self.with_refresh(self.goals.toggle_link, edge_type=Edge.PARENT),
            Qt.Key_L: self.with_refresh(self.goals.toggle_link),
            Qt.Key_Q: self.quit_app.emit,
            Qt.Key_R: self.start_edit('Rename goal', self.goals.rename, self._current_goal_label),
            Qt.Key_V: self.toggle_view,
            Qt.Key_Z: self.toggle_zoom,
            Qt.Key_Escape: self.cancel_edit,
            Qt.Key_Space: self.with_refresh(self.goals.hold_select),
            Qt.Key_Slash: self.show_keys_help,
        }
        if event.key() in key_handlers:
            key_handlers[event.key()]()
        else:
            super().keyPressEvent(event)

    def start_edit(self, label, fn, pre_fn=None):
        def inner():
            self.dockWidget.setWindowTitle(label)
            self.input.setEnabled(True)
            self.input.setFocus(True)
            if pre_fn is not None:
                self.input.setText(pre_fn())
            self.input.returnPressed.connect(self.finish_edit(fn))
            self.cancel.setEnabled(True)
            self.cancel.clicked.connect(self.cancel_edit)
        return inner

    def finish_edit(self, fn):
        def inner():
            self.dockWidget.setWindowTitle('')
            self.input.returnPressed.disconnect()
            fn(self.input.text())
            self.input.setEnabled(False)
            self.input.setText('')
            self.cancel.setEnabled(False)
            self.cancel.clicked.disconnect()
            self.refresh.emit()
        return inner

    def cancel_edit(self):
        self.dockWidget.setWindowTitle('')
        self.input.setEnabled(False)
        self.input.setText('')
        self.cancel.setEnabled(False)
        try:
            self.input.returnPressed.disconnect()
            self.cancel.clicked.disconnect()
        except TypeError:
            pass

    def _current_goal_label(self):
        data = self.goals.q(keys='name,select').values()
        return [x['name'] for x in data if x['select'] == 'select'].pop()

    def with_refresh(self, fn, *args, **kwargs):
        def inner():
            fn(*args, **kwargs)
            self.refresh.emit()
        return inner

    def select_number(self, num):
        def inner():
            self.goals.select(num)
            self.refresh.emit()
        return inner

    def toggle_view(self):
        self.force_refresh = True
        self.goals.next_view()
        self._update_title()
        self.refresh.emit()

    def toggle_zoom(self):
        self.force_refresh = True
        self.goals.toggle_zoom()
        self.refresh.emit()

    def show_keys_help(self):
        self.action_Hotkeys.trigger()

    def show_user_message(self, message):
        self.statusBar().showMessage(message, 10000)


def main(root_script):
    parser = ArgumentParser()
    parser.add_argument('db', nargs='?', default=DEFAULT_DB,
                        help='Path to the database file (sieben.db by default)')
    args = parser.parse_args()
    app = QApplication(sys.argv)
    root = dirname(realpath(root_script))
    w = loadUi(join(root, 'ui', 'main.ui'), SiebenApp(args.db))
    w.about = loadUi(join(root, 'ui', 'about.ui'))
    w.hotkeys = loadUi(join(root, 'ui', 'hotkeys.ui'))
    w.setup()
    w.showMaximized()
    sys.exit(app.exec_())
