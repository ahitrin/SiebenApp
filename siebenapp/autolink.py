from dataclasses import dataclass
from typing import Dict, List, Union, Tuple, Optional, Set

from siebenapp.domain import (
    Graph,
    Command,
    EdgeType,
    ToggleClose,
    Add,
    ToggleLink,
    Insert,
    Rename,
    Delete,
    GoalId,
    RenderResult,
    RenderRow,
)
from siebenapp.goaltree import Goals


@dataclass(frozen=True)
class ToggleAutoLink(Command):
    keyword: str


AutoLinkData = List[Tuple[int, str]]


class AutoLink(Graph):
    def __init__(self, goals: Graph, data: Optional[AutoLinkData] = None):
        super(AutoLink, self).__init__(goals)
        self.keywords: Dict[str, int] = {}
        self.back_kw: Dict[int, str] = {}
        if data:
            for goal_id, keyword in data:
                self.keywords[keyword] = goal_id
                self.back_kw[goal_id] = keyword

    def accept_ToggleAutoLink(self, command: ToggleAutoLink) -> None:
        selected_id: int = self.settings("selection")
        if selected_id in self.goaltree.closed:
            self.error("Autolink cannot be set for closed goals")
            return
        if selected_id == Goals.ROOT_ID:
            self.error("Autolink cannot be set for the root goal")
            return
        keyword: str = command.keyword.lower().strip()
        if selected_id in self.back_kw:
            self.keywords.pop(self.back_kw[selected_id])
            self.back_kw.pop(selected_id)
            self.events().append(("remove_autolink", selected_id))
        if keyword in self.keywords:
            old_id: int = self.keywords.pop(keyword)
            self.back_kw.pop(old_id)
            self.events().append(("remove_autolink", old_id))
        if not keyword:
            # empty keyword? exit right now
            return
        self.keywords[keyword] = selected_id
        self.back_kw[selected_id] = keyword
        self.events().append(("add_autolink", selected_id, keyword))

    def accept_ToggleClose(self, command: ToggleClose) -> None:
        selected_id: int = self.settings("selection")
        if selected_id in self.back_kw:
            self.keywords.pop(self.back_kw[selected_id])
            self.back_kw.pop(selected_id)
            self.events().append(("remove_autolink", selected_id))
        self.goaltree.accept(command)

    def accept_Add(self, command: Add) -> None:
        self._autolink_new_goal(command)

    def accept_Insert(self, command: Insert) -> None:
        self._autolink_new_goal(command)

    def _autolink_new_goal(self, command: Union[Add, Insert]) -> None:
        matching: List[int] = self._find_matching_goals(command.name)
        ids_before: Set[int] = set(self.goaltree.goals.keys())
        self.goaltree.accept(command)
        ids_after: Set[int] = set(self.goaltree.goals.keys())
        ids_diff: Set[int] = ids_after.difference(ids_before)
        if ids_diff:
            added_id: int = ids_diff.pop()
            self._make_links(matching, added_id)

    def accept_Rename(self, command: Rename) -> None:
        matching: List[int] = self._find_matching_goals(command.new_name)
        self.goaltree.accept(command)
        selected_id: int = command.goal_id or self.settings("selection")
        self._make_links(matching, selected_id)

    def accept_Delete(self, command: Delete) -> None:
        selected_id: int = command.goal_id or self.settings("selection")
        edges: Dict[GoalId, List[Tuple[GoalId, EdgeType]]] = {
            row.goal_id: row.edges for row in self.goaltree.q().rows
        }
        goals_to_check: List[int] = [selected_id]
        while goals_to_check:
            goal_id: int = goals_to_check.pop()
            goals_to_check.extend(
                e[0]
                for e in edges[goal_id]
                if e[1] == EdgeType.PARENT and isinstance(e[0], int)
            )
            if goal_id in self.back_kw:
                added_kw: str = self.back_kw.pop(goal_id)
                self.keywords.pop(added_kw)
                self.events().append(("remove_autolink", goal_id))
        self.goaltree.accept(command)

    def _find_matching_goals(self, text: str) -> List[int]:
        return [goal_id for kw, goal_id in self.keywords.items() if kw in text.lower()]

    def _make_links(self, matching_goals: List[int], target_goal: int) -> None:
        if not matching_goals:
            return
        self_children: Dict[GoalId, List[GoalId]] = {
            row.goal_id: [e[0] for e in row.edges] for row in self.goaltree.q().rows
        }
        for add_to in matching_goals:
            if target_goal not in self_children[add_to]:
                self.goaltree.accept(ToggleLink(add_to, target_goal, EdgeType.BLOCKER))

    def q(self) -> RenderResult:
        rows: List[RenderRow] = [
            RenderRow(
                row.goal_id,
                row.raw_id,
                row.name,
                row.is_open,
                row.is_switchable,
                row.select,
                self._add_pseudo_goals(row.edges),
            )
            for row in self.goaltree.q().rows
        ]
        fake_rows: List[RenderRow] = [
            RenderRow(
                AutoLink.fake_id(goal_id),
                -1,
                f"Autolink: '{keyword}'",
                True,
                False,
                None,
                [(goal_id, EdgeType.PARENT)],
            )
            for keyword, goal_id in self.keywords.items()
        ]
        return RenderResult(rows=rows + fake_rows)

    def _add_pseudo_goals(
        self, edges: List[Tuple[GoalId, EdgeType]]
    ) -> List[Tuple[GoalId, EdgeType]]:
        return [
            e if e[0] not in self.keywords.values() else (AutoLink.fake_id(e[0]), e[1])
            for e in edges
        ]

    @staticmethod
    def fake_id(goal_id: GoalId) -> GoalId:
        return -(goal_id + 10) if isinstance(goal_id, int) else goal_id

    @staticmethod
    def export(goals: "AutoLink") -> AutoLinkData:
        return [(goal_id, kw) for goal_id, kw in goals.back_kw.items()]
