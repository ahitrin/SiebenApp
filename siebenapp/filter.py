import collections
from dataclasses import dataclass
from typing import Dict, Any

from siebenapp.domain import Graph, Command, with_key


@dataclass(frozen=True)
class FilterBy(Command):
    """Activate or deactivate filtering"""

    pattern: str


class FilterView(Graph):
    def __init__(self, goaltree: Graph):
        super().__init__(goaltree)
        self.pattern = ""

    def events(self) -> collections.deque:
        return self.goaltree.events()

    def verify(self) -> bool:
        return self.goaltree.verify()

    def settings(self, key: str) -> int:
        return self.goaltree.settings(key)

    def accept_FilterBy(self, event: FilterBy):
        self.pattern = event.pattern.lower()

    @with_key("name")
    @with_key("select")
    def q(self, keys: str = "name") -> Dict[int, Any]:
        unfiltered = self.goaltree.q(keys)
        return self._filter(unfiltered, keys) if self.pattern else unfiltered

    def _filter(self, unfiltered, keys):
        accepted_nodes = [
            goal_id
            for goal_id, attrs in unfiltered.items()
            if self.pattern in attrs["name"].lower()
            or attrs.get("select", None) is not None
        ]
        filtered: Dict[int, Any] = {}
        for goal_id, attrs in unfiltered.items():
            if goal_id in accepted_nodes:
                if "edge" in keys:
                    attrs["edge"] = [e for e in attrs["edge"] if e[0] in accepted_nodes]
                filtered[goal_id] = attrs
        return filtered
