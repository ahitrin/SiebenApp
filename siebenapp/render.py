from collections import defaultdict
from dataclasses import dataclass
from typing import Dict, List, Tuple, Union, Any, Set

from siebenapp.domain import Graph, EdgeType


# GoalId for real nodes is integer: -1, 4, 34, etc
# GoalId for fake nodes (used to build edges) is str: '3_5', '1_1', etc
GoalId = Union[str, int]


@dataclass(frozen=True)
class RenderResult:
    graph: Dict[int, Any]
    edge_opts: Dict[str, Tuple[int, int, int]]


def safe_average(items: List[int]) -> int:
    return int(sum(items) / len(items)) if items else 0


class Renderer:
    WIDTH_LIMIT = 4

    def __init__(self, goals: Graph) -> None:
        self.graph = goals.q(keys="name,edge,open,select,switchable")
        self.edges: Dict[GoalId, List[Any]] = {
            key: [e[0] for e in values["edge"]] for key, values in self.graph.items()
        }
        self.layers: Dict[int, List[GoalId]] = defaultdict(list)
        self.positions: Dict[GoalId, int] = {}
        self.edge_types: Dict[Tuple[GoalId, GoalId], EdgeType] = {
            (parent, child): edge_type
            for parent in self.graph
            for child, edge_type in self.graph[parent]["edge"]
        }
        self.result_edge_options: Dict[str, Tuple[int, int, int]] = {}

    def build(self) -> RenderResult:
        self.split_by_layers()
        self.reorder()
        self.update_graph()
        self.build_index()
        return RenderResult(self.graph, self.result_edge_options)

    def split_by_layers(self) -> None:
        unsorted_goals: Dict[GoalId, List[int]] = dict(self.edges)
        sorted_goals: Set[GoalId] = set()
        incoming_edges: Set[GoalId] = set()
        outgoing_edges: List[GoalId] = []
        current_layer = 0
        while unsorted_goals:
            new_layer: List[GoalId] = []
            for goal, edges_len in self.candidates_for_new_layer(
                sorted_goals, unsorted_goals
            ):
                unsorted_goals.pop(goal)
                sorted_goals.add(goal)
                new_layer.append(goal)
                back_edges = [k for k, vs in self.edges.items() if goal in vs]
                outgoing_edges.extend(iter(back_edges))
                if (len(new_layer) >= self.WIDTH_LIMIT and edges_len < 1) or (
                    len(outgoing_edges) >= self.WIDTH_LIMIT
                ):
                    break
            incoming_edges = incoming_edges.difference(set(new_layer))
            for original_id in incoming_edges:
                new_goal_name = f"{original_id}_{current_layer}"
                self.edges[new_goal_name] = [
                    g
                    for g in self.edges[original_id]
                    if g in sorted_goals and g not in new_layer
                ]
                new_edge_type = EdgeType.BLOCKER
                for g in self.edges[new_goal_name]:
                    self.edges[original_id].remove(g)
                    self.edge_types[new_goal_name, g] = self.edge_types[original_id, g]
                    new_edge_type = max(new_edge_type, self.edge_types[original_id, g])
                self.edges[original_id].append(new_goal_name)
                self.edge_types[original_id, new_goal_name] = new_edge_type
                new_layer.append(new_goal_name)
                sorted_goals.add(new_goal_name)
            self.layers[current_layer] = new_layer
            current_layer += 1
            incoming_edges.update(outgoing_edges)
            outgoing_edges.clear()
        self.positions = {
            g: idx for layer in self.layers.values() for idx, g in enumerate(layer)
        }

    @staticmethod
    def candidates_for_new_layer(
        sorted_goals: Set[GoalId], unsorted_goals: Dict[GoalId, List[int]]
    ) -> List[Tuple[GoalId, int]]:
        candidates = [
            (goal, len(edges))
            for goal, edges in unsorted_goals.items()
            if all(v in sorted_goals for v in edges)
        ]
        candidates.sort(key=lambda x: x[1], reverse=True)
        return candidates

    def reorder(self) -> None:
        for curr_layer in sorted(self.layers.keys(), reverse=True)[:-1]:
            fixed_line = self.layers[curr_layer]
            random_line = self.layers[curr_layer - 1]
            deltas = self.count_deltas(fixed_line)
            new_positions = {
                g: int(self.positions[g] + deltas.get(g, 0)) for g in random_line
            }

            random_line = place(new_positions)
            self.positions.update({g: idx for idx, g in enumerate(random_line)})
            self.layers[curr_layer - 1] = random_line

    def count_deltas(self, fixed_line):
        deltas: Dict[int, List[int]] = defaultdict(list)
        for goal in fixed_line:
            if goal is not None:
                for e in self.edges[goal]:
                    deltas[e].append(self.positions[goal] - self.positions[e])
        return {k: safe_average(v) for k, v in deltas.items()}

    def intersections(self, layer):
        enumerated_edges = [
            (self.positions[t], self.positions[e])
            for t in self.layers[layer]
            for e in self.edges[t]
        ]
        return len(
            [
                1
                for a in enumerated_edges
                for b in enumerated_edges
                if a[0] < b[0] and a[1] > b[1]
            ]
        )

    def update_graph(self):
        for row in sorted(self.layers.keys()):
            real_col = 0
            for col, goal_id in enumerate(self.layers[row]):
                if goal_id is None:
                    continue
                if goal_id in self.graph:
                    self.graph[goal_id]["col1"] = real_col
                    real_col += 1
                else:
                    self.graph[goal_id] = {
                        "name": "",
                        "edge": [],
                        "switchable": False,
                        "select": None,
                        "open": True,
                    }
                self.graph[goal_id].update(
                    {
                        "row": row,
                        "col": col,
                        "edge": [
                            (child, self.edge_types[goal_id, child])
                            for child in self.edges[goal_id]
                        ],
                    }
                )

    def build_index(self):
        result_index: Dict[int, Dict[int, GoalId]] = {}
        for goal_id, attrs in self.graph.items():
            row = attrs["row"]
            col = attrs["col"]
            if row not in result_index:
                result_index[row] = {}
            result_index[row][col] = goal_id

        for row, row_vals in result_index.items():
            left = 0
            edges: List[str] = []
            phase = "goals"

            for col, goal_id in sorted(row_vals.items(), key=lambda x: x[0]):
                if isinstance(goal_id, int):
                    if phase == "edges":
                        self._write_edges(edges, left)
                        edges = []
                    phase = "goals"
                    left = goal_id
                else:
                    phase = "edges"
                    edges.append(goal_id)
            self._write_edges(edges, left)

    def _write_edges(self, edges: List[str], left: int):
        self.result_edge_options.update(
            {e: (left, i + 1, len(edges) + 1) for i, e in enumerate(edges)}
        )


def place(source):
    result = []
    unsorted = sorted(list(source.items()), key=goal_key)
    while unsorted:
        value, index = unsorted.pop(0)
        if len(result) < index + 1:
            result.extend([None] * (index + 1 - len(result)))
        if result[index] is None:
            result[index] = value
        else:
            unsorted.insert(0, (value, index + 1))
    while len(result) > 4 and None in result:
        result.remove(None)
    return result


def goal_key(tup):
    """Sort goals by position first and by id second (transform str ids into ints)"""
    goal_id, goal_pos = tup
    if isinstance(goal_id, str):
        return goal_pos, int(goal_id.split("_")[0])
    return goal_pos, goal_id
