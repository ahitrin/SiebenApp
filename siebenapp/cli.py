import sys
from argparse import ArgumentParser
from operator import itemgetter

from siebenapp.render import Renderer
from siebenapp.domain import (
    ToggleClose,
    Delete,
    HoldSelect,
    ToggleLink,
    EdgeType,
    Select,
    Add,
    Insert,
    Rename,
    Graph,
)
from siebenapp.progress_view import ToggleProgress
from siebenapp.filter_view import FilterBy
from siebenapp.open_view import ToggleOpenView
from siebenapp.switchable_view import ToggleSwitchableView
from siebenapp.system import save, load
from siebenapp.zoom import ToggleZoom

USER_MESSAGE = ""


class IO:
    """Simple wrapper for all input/output interactions"""

    def write(self, text: str, *args) -> None:
        raise NotImplementedError

    def read(self) -> str:
        raise NotImplementedError


class ConsoleIO(IO):
    def __init__(self, prompt):
        super().__init__()
        sys.stdin.reconfigure(encoding="utf-8")
        self.prompt = prompt

    def write(self, text: str, *args) -> None:
        print(text, *args)

    def read(self) -> str:
        return input(self.prompt)


def update_message(message=""):
    global USER_MESSAGE
    USER_MESSAGE = message


def fmt(goal_id, goal_vars, id_width):
    if goal_id < 0:
        goal_id = " " * id_width
    name = goal_vars["name"]
    op = " " if goal_vars["open"] else "x"
    status = f"[{op}]" if goal_vars["switchable"] else f" {op} "
    selection = " "
    if goal_vars["select"] == "select":
        selection = ">"
    elif goal_vars["select"] == "prev":
        selection = "_"
    children = ""
    if goal_vars["edge"]:
        child_list = ",".join(str(e[0]) for e in goal_vars["edge"])
        children = f" [{child_list}]"
    return f"{goal_id}{status}{selection}{name}{children}"


def build_actions(command):
    simple_commands = {
        "c": ToggleClose(),
        "d": Delete(),
        "h": HoldSelect(),
        "k": ToggleLink(edge_type=EdgeType.PARENT),
        "l": ToggleLink(),
        "n": ToggleOpenView(),
        "p": ToggleProgress(),
        "t": ToggleSwitchableView(),
        "z": ToggleZoom(),
    }
    if command and all(c in "1234567890" for c in command):
        return [Select(int(c)) for c in command]
    if command.startswith("a "):
        return [Add(command[2:])]
    if command.startswith("i "):
        return [Insert(command[2:])]
    if command.startswith("r "):
        return [Rename(command[2:])]
    if command.startswith("f"):
        return [FilterBy(command[1:].lstrip())]
    if command in simple_commands:
        return [simple_commands[command]]
    return []


def loop(io: IO, goals: Graph, db_name: str):
    cmd = ""
    while cmd != "q":
        render_result = Renderer(goals, 100).build()
        rows = render_result.graph
        index = sorted(
            [
                (goal_id, goal_vars["row"], goal_vars["col"])
                for goal_id, goal_vars in rows.items()
                if isinstance(goal_id, int)
            ],
            key=itemgetter(1, 2),
            reverse=True,
        )
        id_width = len(str(max(g for g in rows.keys() if isinstance(g, int))))
        for item in index:
            goal_id = item[0]
            goal_vars = rows[goal_id]
            io.write(fmt(goal_id, goal_vars, id_width))
        if USER_MESSAGE:
            io.write(USER_MESSAGE)
        update_message()
        try:
            cmd = io.read().strip()
        except EOFError:
            break
        actions = build_actions(cmd)
        if actions:
            goals.accept_all(*actions)
        save(goals, db_name)


def main():
    io = ConsoleIO(prompt="> ")
    parser = ArgumentParser()
    parser.add_argument(
        "db",
        nargs="?",
        default="sieben.db",
        help="Path to the database file",
    )
    args = parser.parse_args()
    goals = load(args.db, update_message)
    loop(io, goals, args.db)
