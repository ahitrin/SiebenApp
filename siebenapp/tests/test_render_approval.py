import io
from pprint import pprint

from approvaltests import verify  # type: ignore
from approvaltests.reporters import GenericDiffReporterFactory  # type: ignore

from siebenapp.render import (
    Renderer,
    GeometryProvider,
    render_lines,
    Point,
    adjust_graph,
)
from siebenapp.tests.dsl import build_goaltree, open_, clos_, selected, previous


class FakeGeometry(GeometryProvider):
    def width(self, row, col):
        return Point(44 + 6 * (row % 2), 0)

    def height(self, row, col):
        return Point(0, 40 + 10 * (col % 2))

    def top_left(self, row, col):
        return Point(col * 100, row * 100)

    def top_right(self, row, col):
        return self.top_left(row, col) + self.width(row, col)

    def bottom_left(self, row, col):
        return self.top_left(row, col) + self.height(row, col)

    def bottom_right(self, row, col):
        return self.bottom_left(row, col) + self.width(row, col)


def test_render_example():
    g = build_goaltree(
        open_(1, "Root", [2, 3, 4, 5, 6], [7, 8]),
        clos_(2, "Closed", blockers=[7]),
        open_(3, "Simply 3", blockers=[5, 8]),
        open_(4, "Also 4", blockers=[5, 6, 7, 8], select=selected),
        open_(5, "Now 5", blockers=[6]),
        clos_(6, "Same 6", blockers=[7]),
        clos_(7, "Lucky 7", [8], select=previous),
        clos_(8, "Finally 8"),
    )
    r = Renderer(g)
    result = r.build()
    gp = FakeGeometry()
    adjust_graph(result, gp)
    lines = render_lines(gp, result)
    with io.StringIO() as out:
        print("== Graph\n", file=out)
        pprint(result.graph, out)
        print("\n== Geometry change after adjust\n", file=out)
        total_delta = Point(0, 0)
        for goal_id in g.goals:
            goal = result.graph[goal_id]
            delta = gp.top_left(goal["row"], goal["col"]) - Point(goal["x"], goal["y"])
            total_delta += delta
            print(f"{goal_id}: dx={delta.x}, dy={delta.y}", file=out)
        avg_dx = total_delta.x // len(g.goals)
        avg_dy = total_delta.y // len(g.goals)
        print(f"Avg: dx={avg_dx}, dy={avg_dy}", file=out)
        print("\n== Edge options\n", file=out)
        pprint(result.edge_opts, out)
        print("\n== Lines", file=out)
        pprint(lines, out)
        verify(
            out.getvalue(), reporter=GenericDiffReporterFactory().get_first_working()
        )
