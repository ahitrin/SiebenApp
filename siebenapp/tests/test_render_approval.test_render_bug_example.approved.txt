== Graph

{1: {'col': 0,
     'edge': [(3, <EdgeType.PARENT: 2>),
              (4, <EdgeType.PARENT: 2>),
              ('1_4', <EdgeType.PARENT: 2>)],
     'name': 'Root',
     'open': True,
     'row': 5,
     'select': None,
     'switchable': False},
 2: {'col': 1,
     'edge': [(7, <EdgeType.BLOCKER: 1>)],
     'name': 'Closed',
     'open': False,
     'row': 2,
     'select': None,
     'switchable': True},
 3: {'col': 1,
     'edge': [(5, <EdgeType.BLOCKER: 1>), ('3_3', <EdgeType.BLOCKER: 1>)],
     'name': 'Simply 3',
     'open': True,
     'row': 4,
     'select': None,
     'switchable': False},
 4: {'col': 2,
     'edge': [(5, <EdgeType.BLOCKER: 1>), ('4_3', <EdgeType.BLOCKER: 1>)],
     'name': 'Also 4',
     'open': True,
     'row': 4,
     'select': 'select',
     'switchable': False},
 5: {'col': 2,
     'edge': [(6, <EdgeType.BLOCKER: 1>)],
     'name': 'Now 5',
     'open': True,
     'row': 3,
     'select': None,
     'switchable': True},
 6: {'col': 3,
     'edge': [(7, <EdgeType.BLOCKER: 1>)],
     'name': 'Same 6',
     'open': False,
     'row': 2,
     'select': None,
     'switchable': True},
 7: {'col': 2,
     'edge': [(8, <EdgeType.PARENT: 2>)],
     'name': 'Lucky 7',
     'open': False,
     'row': 1,
     'select': 'prev',
     'switchable': True},
 8: {'col': 1,
     'edge': [],
     'name': 'Finally 8',
     'open': False,
     'row': 0,
     'select': None,
     'switchable': True},
 '1_1': {'col': 0,
         'edge': [(8, <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 1,
         'select': None,
         'switchable': False},
 '1_2': {'col': 0,
         'edge': [(7, <EdgeType.BLOCKER: 1>), ('1_1', <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 2,
         'select': None,
         'switchable': False},
 '1_3': {'col': 0,
         'edge': [(2, <EdgeType.PARENT: 2>),
                  (6, <EdgeType.PARENT: 2>),
                  ('1_2', <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 3,
         'select': None,
         'switchable': False},
 '1_4': {'col': 0,
         'edge': [(5, <EdgeType.PARENT: 2>), ('1_3', <EdgeType.PARENT: 2>)],
         'name': '',
         'open': True,
         'row': 4,
         'select': None,
         'switchable': False},
 '3_1': {'col': 1,
         'edge': [(8, <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 1,
         'select': None,
         'switchable': False},
 '3_2': {'col': 2,
         'edge': [('3_1', <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 2,
         'select': None,
         'switchable': False},
 '3_3': {'col': 1,
         'edge': [('3_2', <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 3,
         'select': None,
         'switchable': False},
 '4_1': {'col': 3,
         'edge': [(8, <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 1,
         'select': None,
         'switchable': False},
 '4_2': {'col': 4,
         'edge': [(7, <EdgeType.BLOCKER: 1>), ('4_1', <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 2,
         'select': None,
         'switchable': False},
 '4_3': {'col': 3,
         'edge': [(6, <EdgeType.BLOCKER: 1>), ('4_2', <EdgeType.BLOCKER: 1>)],
         'name': '',
         'open': True,
         'row': 3,
         'select': None,
         'switchable': False}}

== Layers

{}
